controllers/service_impl.go

```diff
@@ -0,0 +1,240 @@
+func splitApplicationKey(applicationKey string) (string, string) {
+	var name, namespace string
+	names := strings.Split(applicationKey, "/")
+	if len(names) > 1 {
+		namespace, name = names[0], names[1]
+	} else {
+		namespace, name = "default", names[0]
+	}
+
+	return namespace, name
+}
+
+//
+//
+//
+
+// writeMetricInternal is detailed implementation of gRPC method WriteMetric
+func writeMetricInternal(application, user, metric, valueStr string, client k8sclient.Interface) error {
+	log.Logger.Tracef("writeMetricInternal called for application, user: %s, %s", application, user)
+	defer log.Logger.Trace("writeMetricInternal completed")
+
+	s, track, err := lookupInternal(application, user)
+	if err != nil || track == nil {
+		return err
+	}
+	log.Logger.Debug(fmt.Sprintf("lookup(%s,%s) -> %d", application, user, *track))
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if s.Versions[*track].Metrics == nil {
+		s.Versions[*track].Metrics = map[string]*summarymetrics.SummaryMetric{}
+	}
+	m, ok := s.Versions[*track].Metrics[metric]
+	if !ok {
+		m = summarymetrics.EmptySummaryMetric()
+		s.Versions[*track].Metrics[metric] = m
+	}
+
+	value, err := strconv.ParseFloat(valueStr, 64)
+	if err != nil {
+		log.Logger.Warn("Unable to parse metric value ", valueStr)
+		return err
+	}
+	m.Add(value)
+
+	// persist updated metric
+	legacyApp := routemapToLegacyApplication(s)
+	return write(client, legacyApp)
+}
+
+func routemapToLegacyApplication(s *routemap) abn.LegacyApplication {
+
+	tracks := make(abn.LegacyTracks, len(s.Versions))
+	versions := make(abn.LegacyVersions, len(s.Versions))
+	for t, v := range s.Versions {
+		asStr := fmt.Sprintf("%d", t)
+		tracks[asStr] = asStr
+		versions[asStr] = &abn.LegacyVersion{
+			Metrics: v.Metrics,
+		}
+	}
+
+	a := abn.LegacyApplication{
+		Name:     s.Namespace + "/" + s.Name,
+		Tracks:   tracks,
+		Versions: versions,
+	}
+	return a
+}
+
+const secretKey string = "application.yaml"
+
+func write(client k8sclient.Interface, a abn.LegacyApplication) error {
+	var secret *corev1.Secret
+
+	// marshal to byte array
+	// note that this uses a custom MarshalJSON that removes untracked
+	// versions if the application data is too large
+	rawData, err := yaml.Marshal(a)
+	if err != nil {
+		return err
+	}
+
+	secretNamespace := namespaceFromKey(a.Name)
+	secretName := nameFromKey(a.Name)
+
+	// get the current secret; it will have been created as part of install
+	secret, err = client.GetSecret(secretNamespace, secretName)
+	if err != nil {
+		log.Logger.Error("secret does not exist; no metrics can be recorded")
+		return err
+	}
+
+	if secret.Data == nil {
+		secret.Data = map[string][]byte{}
+	}
+
+	secret.Data[secretKey] = rawData
+	if secret.StringData != nil {
+		secret.StringData[secretKey] = string(rawData)
+	}
+
+	// update the secret
+	_, err = client.UpdateSecret(secret)
+	if err != nil {
+		log.Logger.WithError(err).Warn("unable to persist app data")
+		return err
+	}
+
+	return nil
+}
+
+//
+//
+//
```

Comment:
Are these intentional?

Reply:
they were intentional to separate the implementation of the different APIs. Not strictly required.